diff --git a/src/lj_api.c b/src/lj_api.c
index e9fc25b4..4559cdb5 100644
--- a/src/lj_api.c
+++ b/src/lj_api.c
@@ -9,6 +9,8 @@
 #define lj_api_c
 #define LUA_CORE
 
+#include <stdatomic.h>
+
 #include "lj_obj.h"
 #include "lj_gc.h"
 #include "lj_err.h"
@@ -1317,3 +1319,85 @@ LUA_API void lua_setallocf(lua_State *L, lua_Alloc f, void *ud)
   g->allocf = f;
 }
 
+/*------------*/
+/* Extensions */
+/*------------*/
+
+LUALIB_API int lua_ext_tab_len(lua_State* L, int idx, MSize* outsize)
+{
+    TValue* o = index2adr(L, idx);
+    GCtab* tab = tabV(o);
+    if (LJ_LIKELY(tab->hmask == 0))
+        return lj_tab_len(tab);
+    // Return an error as the table has a hash part.
+    return -1;
+}
+
+LUALIB_API lua_Number lua_ext_fast_checknumber(lua_State *L, int idx)
+{
+  cTValue *o = index2adr(L, idx);
+  if (LJ_LIKELY(tvisnumber(o)))
+    return numberVnum(o);
+  else
+    lj_err_argt(L, idx, LUA_TNUMBER);
+  return 0;
+}
+
+LUALIB_API lua_Integer lua_ext_fast_checkinteger(lua_State *L, int idx)
+{
+  cTValue *o = index2adr(L, idx);
+  lua_Number n;
+  if (LJ_LIKELY(tvisint(o))) {
+    return intV(o);
+  } else if (LJ_LIKELY(tvisnum(o))) {
+    n = numV(o);
+  } else {
+    lj_err_argt(L, idx, LUA_TNUMBER);
+  }
+#if LJ_64
+  return (lua_Integer)n;
+#else
+  return lj_num2int(n);
+#endif
+}
+
+LUALIB_API int lua_ext_fast_checkboolean(lua_State *L, int idx)
+{
+  cTValue *o = index2adr(L, idx);
+  if (LJ_LIKELY(tvisbool(o))) {
+    return (int)intV(o);
+  } else {
+    lj_err_argt(L, idx, LUA_TNUMBER);
+  }
+}
+
+static atomic_intptr_t BP3D_LUA_EXT_KEYREG = 0;
+static atomic_uint_fast32_t BP3D_LUA_EXT_KEYREG_REFS = 0;
+
+LUALIB_API void* lua_ext_keyreg_get() {
+    return (void*)BP3D_LUA_EXT_KEYREG;
+}
+
+LUALIB_API void* lua_ext_keyreg_ref(void* ptr) {
+    int flag = 0;
+    if (BP3D_LUA_EXT_KEYREG_REFS == 0) {
+        BP3D_LUA_EXT_KEYREG = (atomic_intptr_t)ptr;
+        flag = 1;
+    }
+    (unsigned int)++BP3D_LUA_EXT_KEYREG_REFS;
+    if (flag == 1) {
+        return NULL;
+    } else {
+        return ptr;
+    }
+}
+
+LUALIB_API void* lua_ext_keyreg_unref() {
+    unsigned int refs = (unsigned int)--BP3D_LUA_EXT_KEYREG_REFS;
+    if (refs == 0) {
+        void* ptr = (void*)BP3D_LUA_EXT_KEYREG;
+        BP3D_LUA_EXT_KEYREG = 0;
+        return ptr;
+    }
+    return NULL;
+}
