diff --git a/src/msvcbuild.bat b/src/msvcbuild.bat
index 69c0c61a..0665eaa3 100644
--- a/src/msvcbuild.bat
+++ b/src/msvcbuild.bat
@@ -29,8 +29,6 @@
 @set DASMDIR=..\dynasm
 @set DASM=%DASMDIR%\dynasm.lua
 @set DASC=vm_x64.dasc
-@set LJDLLNAME=lua51.dll
-@set LJLIBNAME=lua51.lib
 @set ALL_LIB=lib_base.c lib_math.c lib_bit.c lib_string.c lib_table.c lib_io.c lib_os.c lib_package.c lib_debug.c lib_jit.c lib_ffi.c lib_buffer.c
 
 @setlocal
